### PR TITLE
[Fixes #62] Added google authentication

### LINKED ISSUE #62: Add Social Media authentication
We can add authentication by various social networking handles like gmail, facebook etc.

### PR DESCRIPTION
**Pull request by Vrindagupta6828**
Work in progress
See #62

### CODE REVIEW

**Comment by Vrindagupta6828** [comment]
@garg3133 Please give me some feedback as well as help to solve this issue. Error faced is:
Capture

**Comment by garg3133** [comment]
You may need to change the following block of code to generate a random alphanumeric password if no password is provided while creating a new user.

https://github.com/garg3133/JagratiWebApp/blob/c993b2ba7fe5423da780682f125c286511baae11/accounts/models.py#L17

Also, put all the settings you are adding to `base.py` files and not `development.py` so that they can also be used in production.

**Comment by Vrindagupta6828** [comment]
@garg3133 The issue is resolved thanks a lot for your guidance.
Please review my PR.

File: Jagrati/settings/base.py [hunk 1]
```diff
@@ -118,3 +119,12 @@
 
 
 AUTH_USER_MODEL = 'accounts.User'
+
+AUTHENTICATION_BACKENDS=['django.contrib.auth.backends.ModelBackend',
```
Comment (garg3133) [inline Jagrati/settings/base.py:123]: Put the `AUTHENTICATION_BACKENDS` after `WSGI_APPLICATION` and with proper indentation like other variables.

File: Jagrati/settings/base.py [hunk 2]
```diff
@@ -118,3 +119,12 @@
...
+                         'social_core.backends.google.GoogleOAuth2']
+
+ALLOWED_HOSTS=['jagrati.herokuapp.com','localhost','127.0.0.1']
```
Comment (garg3133) [inline Jagrati/settings/base.py:126]: We don't need this line as for development, allowed hosts are already specified in `development.py` and for production, we fetch allowed hosts from environment variables.

File: Jagrati/settings/base.py [hunk 3]
```diff
@@ -118,3 +119,12 @@
...
+
+SOCIAL_AUTH_GOOGLE_OAUTH2_KEY = '108126785604-k1sngb8ohsodee9rdvq5k5q9t3tss8k6.apps.googleusercontent.com'
+SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET = 'RvEINK8clQL0ld4tBZxXalf8'
```
Comment (garg3133) [inline Jagrati/settings/base.py:129]: You should never put these keys open in settings. Instead, put these keys `.env` file and call the values from env variables using `config` (See `production.py` for how to use).

File: accounts/models.py [hunk 4]
```diff
@@ -15,7 +16,8 @@ def create_user(self, email, password=None):  # Add all required fields
         if not email:
             raise ValueError("User must have an email address")
         if not password:
-            raise ValueError("User must have a password")
+            password=get_random_string(length=10);
+            #raise ValueError("User must have a password")
```
Comment (garg3133) [inline accounts/models.py:20]: Remove `;`, add `space` around `=` and remove the commented line.

File: templates/accounts/login_signup.html [hunk 5]
```diff
@@ -60,6 +60,11 @@ <h1>Welcome Back!</h1>
                     <h1>Hello, Friend!</h1>
                     <p>Enter your personal details and start journey with us</p>
                     <button class="ghost" id="signUp">Sign Up</button>
+                    <div class="social">
+                        <ul>
+                            <button type='submit'><a href="{% url "social:begin" "google-oauth2" %}?next={{ request.path }}">Sign in with Google</a></button>
+                        </ul>
+                    </div>
```
Comment (garg3133) [inline templates/accounts/login_signup.html:67]: The position and style of social auth button doesn't look good here. Instead, put the button in between Sign in and email field with some design like the following:

Screenshot from 2021-01-27 23-26-48

**Review by garg3133 — COMMENTED** [review]
Also, you've not added the context processors and middleware for `social_auth`. See https://simpleisbetterthancomplex.com/tutorial/2016/10/24/how-to-add-social-login-to-django.html

**Comment by Vrindagupta6828** [comment]
@garg3133 I have resolved all errors please review my PR.

**Comment by garg3133** [comment]
The bootstrap files you added to login_signup page had broke the design of the whole page. But then I thought maybe you won't be able to achieve the exact design I wanted for the button. So, I made it myself.

Apart from that, whenever you install a new package in the app, always run `pip freeze > requirements.txt` so that all the dependencies get added to the requirements file.

Rest all the changes looked good. Merged this PR into master after making the above said changes.

Thank you for your contribution :tada:

**Pull request merged**